codegenerator/cli/templates/static/codegen/src/TopicFilter.res

```diff
@@ -0,0 +1,27 @@
+let toTwosComplement = (num: bigint, ~bytesLen: int) => {
+  let maxValue = 1n->BigInt.Bitwise.shift_left(BigInt.fromInt(bytesLen * 8))
+  let mask = maxValue->BigInt.sub(1n)
+  num->BigInt.add(maxValue)->BigInt.Bitwise.logand(mask)
+}
+
+let fromSignedBigInt = val => {
+  let bytesLen = 32
+  let val = val >= 0n ? val : val->toTwosComplement(~bytesLen)
+  val->Viem.bigintToHex(~options={size: bytesLen})
+}
+
+type hex = EvmTypes.Hex.t
+//bytes currently does not work with genType and we also currently generate bytes as a string type
+type bytesHex = string
+let keccak256 = Viem.keccak256
+let bytesToHex = Viem.bytesToHex
+let concat = Viem.concat
+let castToHexUnsafe: 'a => hex = val => val->Utils.magic
+let fromBigInt: bigint => hex = val => val->Viem.bigintToHex(~options={size: 32})
+let fromDynamicString: string => hex = val => val->(Utils.magic: string => hex)->keccak256
+let fromString: string => hex = val => val->Viem.stringToHex(~options={size: 32})
+let fromAddress: Address.t => hex = addr => addr->(Utils.magic: Address.t => hex)->Viem.pad
+let fromDynamicBytes: bytesHex => hex = bytes => bytes->(Utils.magic: bytesHex => hex)->keccak256
+let fromBytes: bytesHex => hex = bytes =>
+  bytes->(Utils.magic: bytesHex => bytes)->Viem.bytesToHex(~options={size: 32})
+let fromBool: bool => hex = b => b->Viem.boolToHex(~options={size: 32})
```

Comment:
A collection of TopicFilter functions that are combined when generating topic filters for different types

Reply:
Maybe we can simplify it and do use something like this instead https://viem.sh/docs/contract/encodeEventTopics.html ?

Also, I'd like to move as much logic as possible from codegen. So doing encoding with Viem instead of selecting the right encoder in Rust, will make the envio closer to the library-like state

Reply:
It's not a bad approach. I'm not sure of it's limitations though. I  would still need to generate each of those selectors. And it doesn't line up exactly with what we are supporting with our API. We're allowing multiple sets of topic1-3. 

I personally prefer this as is right now. I can swap out my own hex encoding functions from viem or use another library and everthing works. No need to rely on viem to give as a full implementation.

I think for user API we still want to control that with codegen. For a library implementation I think they would just encode the topics themselves and pass it in.